File: backend/api/update.py
"""更新检查与执行 API"""
import logging
import sys

from fastapi import APIRouter

from core.config import _get_exe_dir
from core.update.checker import check_for_update
from core.update.updater import perform_update
from core.version import __version__

router = APIRouter()
logger = logging.getLogger(__name__)


def get_install_mode() -> str:
    """检测当前安装模式: dev / setup / portable / naked"""
    if not getattr(sys, "frozen", False):
        return "dev"
    exe_dir = _get_exe_dir()
    if exe_dir is None:
        return "naked"
    if (exe_dir / "portable.flag").exists():
        return "portable"
    if (exe_dir / "installed.flag").exists():
        return "setup"
    return "naked"


@router.get("/api/check_update", summary="检查更新")
async def check_update():
    """检查 GitHub Releases 是否有新版本"""
    result = check_for_update(__version__)
    result["install_mode"] = get_install_mode()
    return result


@router.post("/api/perform_update", summary="执行更新")
async def do_update():
    """下载并执行更新（仅安装版/便携版/裸exe可用）"""
    mode = get_install_mode()
    if mode == "dev":
        return {"ok": False, "message": "开发模式不支持自动更新"}

    check_result = check_for_update(__version__)
    if not check_result.get("update_available"):
        return {"ok": False, "message": "当前已是最新版本"}

    download_urls = check_result.get("download_urls", {})
    return perform_update(mode, download_urls)

File: frontend/src-tauri/src/commands/update.rs
use tauri::AppHandle;
use tauri_plugin_updater::UpdaterExt;

use crate::error::{AppError, AppResult};
use crate::update::{self, CheckUpdateResponse, PerformUpdateResponse};

/// Check for updates via the GitHub Releases API (for display: version + release notes).
///
/// The actual install path is handled by the native `tauri-plugin-updater` in `do_update`.
#[tauri::command(rename_all = "snake_case")]
pub async fn check_update(app: AppHandle) -> AppResult<CheckUpdateResponse> {
    let current_version = app.package_info().version.to_string();
    let install_mode = detect_install_mode();

    let result = tokio_or_spawn(move || update::check_for_update(&current_version, &install_mode));

    Ok(result)
}

/// Download and install the latest update via the native `tauri-plugin-updater`.
///
/// Requires release infrastructure: a signed `latest.json` manifest hosted at the
/// configured `plugins.updater.endpoints` URL and the matching public key in
/// `plugins.updater.pubkey`. Until that is set up, this returns a graceful error.
#[tauri::command(rename_all = "snake_case")]
pub async fn do_update(app: AppHandle) -> AppResult<PerformUpdateResponse> {
    let updater = app
        .updater()
        .map_err(|e| {
            log::warn!("[UPDATE_ERROR] do_update: updater not configured | {}", e);
            AppError::UpdateError(format!("updater 未配置: {e}"))
        })?;

    let update = updater
        .check()
        .await
        .map_err(|e| {
            log::warn!("[UPDATE_ERROR] do_update: check failed | {}", e);
            AppError::UpdateError(format!("检查更新失败: {e}"))
        })?;

    let Some(update) = update else {
        return Ok(PerformUpdateResponse {
            ok: false,
            message: "当前已是最新版本".to_string(),
        });
    };

    update
        .download_and_install(|_chunk, _total| {}, || {})
        .await
        .map_err(|e| {
            log::error!("[UPDATE_ERROR] do_update: download_and_install failed | version={} {}", update.version, e);
            AppError::UpdateError(format!("下载/安装更新失败: {e}"))
        })?;

    app.restart()
}

fn detect_install_mode() -> String {
    // Simple heuristic: check if running from a portable directory or installed location
    if let Ok(exe) = std::env::current_exe() {
        let exe_dir = exe.parent().unwrap_or(std::path::Path::new("."));
        let portable_flag = exe_dir.join("portable.flag");
        if portable_flag.exists() {
            return "portable".to_string();
        }
        // Check if in Program Files (installed via setup)
        let exe_str = exe.to_string_lossy().to_lowercase();
        if exe_str.contains("program files") {
            return "setup".to_string();
        }
    }
    "dev".to_string()
}

/// Run a blocking closure on a separate thread and wait for the result.
fn tokio_or_spawn<F, T>(f: F) -> T
where
    F: FnOnce() -> T + Send + 'static,
    T: Send + 'static,
{
    std::thread::spawn(f)
        .join()
        .expect("background thread panicked")
}
